### PR TITLE
Fixes #11317 - Support basic auth for external sources

### PR DESCRIPTION
**Pull request by ofedoren**
Requires:
- [x] - https://github.com/Apipie/apipie-bindings/pull/82
- [x] - https://github.com/theforeman/puppet-foreman/pull/1108

### CODE REVIEW

**Review by adamruzicka — COMMENTED** [review]
Tested this together with apipie-bindings and foreman-puppet changes, works well. Error handling in case the authentication fails could use some polish though

File: lib/hammer_cli_foreman/api/basic_auth.rb [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+      def error(ex)
+        return unless ex.is_a?(RestClient::Unauthorized)
+
+        clear
+        message = _('Invalid username or password.')
+        begin
+          message = JSON.parse(ex.response.body)['error']['message']
+        rescue
+        end
+        UnauthorizedError.new(message)
```
[the remark this answers is not in the document]
Comment (ofedoren) [reply lib/hammer_cli_foreman/api/basic_auth.rb]: Thanks, updated. It should show `Invalid credentials` error now.

File: lib/hammer_cli_foreman/api/basic_auth.rb [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+      def error(ex)
+        return unless ex.is_a?(RestClient::Unauthorized)
+
+        clear
+        message = _('Invalid username or password.')
+        begin
+          response_msg = JSON.parse(ex.response.body)['error']
+        rescue
+        end
+        UnauthorizedError.new(response_msg&.dig('message') || message)
```
Comment (adamruzicka) [inline lib/hammer_cli_foreman/api/basic_auth.rb:21]:
```shell
# hammer -d auth login basic-external -u user -p badpass
----B<-----SNIP----B<-----
[DEBUG 2023-02-10T11:12:49 API] Using authenticator: HammerCLIForeman::Api::SessionAuthenticatorWrapper
[ERROR 2023-02-10T11:12:49 API] 401 Unauthorized
[DEBUG 2023-02-10T11:12:49 API] {
    "error" => "External authentication did not pass."
}
[DEBUG 2023-02-10T11:12:49 Exception] Using exception handler HammerCLIForeman::ExceptionHandler#handle_general_exception
[ERROR 2023-02-10T11:12:49 Exception] Error: undefined method `dig' for "External authentication did not pass.":String
Error: undefined method `dig' for "External authentication did not pass.":String
[ERROR 2023-02-10T11:12:49 Exception]

NoMethodError (undefined method `dig' for "External authentication did not pass.":String):
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/api/basic_auth.rb:21:in `error'
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/api/session_authenticator_wrapper.rb:66:in `error'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/api.rb:248:in `rescue in http_call'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/api.rb:238:in `http_call'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/api.rb:190:in `call_action'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/api.rb:185:in `call'
    /usr/share/gems/gems/apipie-bindings-0.5.0/lib/apipie_bindings/action.rb:14:in `call'
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/api/connection.rb:36:in `login'
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/authenticate/login.rb:12:in `execute_with_params'
    /usr/share/gems/gems/hammer_cli_foreman-3.6.0.pre.develop/lib/hammer_cli_foreman/auth.rb:41:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/subcommand/execution.rb:11:in `execute'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:63:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/lib/hammer_cli/abstract.rb:103:in `run'
    /usr/share/gems/gems/clamp-1.1.2/lib/clamp/command.rb:132:in `run'
    /usr/share/gems/gems/hammer_cli-3.6.0.pre.develop/bin/hammer:147:in `<top (required)>'
    /usr/bin/hammer:23:in `load'
    /usr/bin/hammer:23:in `<main>'
```
```suggestion
        message = _('Invalid username or password.')
        begin
          response_msg = JSON.parse(ex.response.body)
        rescue
        end
        UnauthorizedError.new(response_msg&.dig('error') || message)
```
```

**Review by adamruzicka — CHANGES_REQUESTED** [review]
Happy path works, non happy still has some issues

File: lib/hammer_cli_foreman/api/basic_auth.rb (hunk 2 above)
Comment (ofedoren) [reply lib/hammer_cli_foreman/api/basic_auth.rb:21]: Right :/ I'm going to rewrite it a bit. The problem is that we don't have consistent response here :/ `JSON.parse(ex.response.body)['error']` can be either string or a hash with `message` entry.

Comment (ofedoren) [reply lib/hammer_cli_foreman/api/basic_auth.rb]: I've changed the message. It now should show the response from the server (if any) plus the default message about invalid credentials (which is most obvious reason to fail for basic auth).

**Review by adamruzicka — APPROVED** [review]
Works well, ACK

**Comment by ofedoren** [comment]
I'll wait for https://github.com/theforeman/puppet-foreman/pull/1108 to be merged and after I'll merge this one. `apipie-bindings` part was merged and released.

**Comment by adamruzicka** [comment]
Could we get a apipie dependency version bump?

**Comment by ofedoren** [comment]
> Could we get a apipie dependency version bump?

I've already released apipie-bindings, hammer is pinned to the latest available version already, that's why the tests are passing now :)

**Comment by adamruzicka** [comment]
> Or do you mean hammer-cli should update minimal version of apipie-bindings?

That's what I meant, or is it fine to treat 0.6.0 as optional?

**Comment by ofedoren** [comment]
> > Or do you mean hammer-cli should update minimal version of apipie-bindings?
> 
> That's what I meant, or is it fine to treat 0.6.0 as optional?

Nah... newer version of hammer-cli-foreman will be broken with old version of apipie :/ Here: https://github.com/theforeman/hammer-cli/pull/363

**Comment by adamruzicka** [comment]
Alright, let's go

**Comment by adamruzicka** [comment]
Thank you @ofedoren !

**Pull request merged**